File: Leetcode-Problems/14 - LongestCommonPrefix.js
//Write a function to find the longest common prefix string amongst an array of strings.
//If there is no common prefix, return an empty string "".
//
//Example 1:
//Input: strs = ["flower","flow","flight"]
//Output: "fl"

//Solution:
const longestCommonPrefix = (strs) => {
    // Edge case: If the input array is empty, return an empty string
    if (!strs.length) return "";

    // Edge case: If there's only one string, it is the common prefix
    if (strs.length === 1) return strs[0];

    // Initialize the prefix with the first string in the array
    let pre = strs[0];

    // Iterate through the array of strings starting from the second string
    for (let i = 1; i < strs.length; i++) {
        // Reduce the prefix until it matches the beginning of the current string
        while (strs[i].indexOf(pre) !== 0) {
            // Remove the last character from the prefix
            pre = pre.slice(0, pre.length - 1);

            // If the prefix becomes empty, no common prefix exists
            if (!pre) return "";
        }
    }

    // Return the longest common prefix found
    return pre;
};

File: Platforms/LeetCode/14-LongestCommonPrefix.cpp
// ✅ Time Complexity: O(N * M)
// N = number of strings in the vector
// M = length of the shortest string
// Because for each character in the first string, we compare with all others.

// ✅ Space Complexity: O(1) (excluding the output string)

#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    string longestCommonPrefix(vector<string> &strs)
    {
        string res = "";

        // Edge case: empty input
        if (strs.empty())
            return res;

        // Loop through characters of the first string
        for (int i = 0; i < strs[0].size(); i++)
        {
            // Compare this character with the same position in all other strings
            for (int j = 0; j < strs.size() - 1; j++)
            {
                // Check if index is out of range or characters differ
                if (i >= strs[j].size() || i >= strs[j + 1].size() || strs[j][i] != strs[j + 1][i])
                    return res;
            }

            // If all matched, add to result
            res += strs[0][i];
        }

        return res;
    }
};

int main()
{
    Solution sol;
    vector<string> strs = {"flower", "flow", "flight"};
    string result = sol.longestCommonPrefix(strs);
    cout << "Longest Common Prefix: " << result << endl;

    return 0;
}
